src/utils/lazy.ts

```diff
@@ -7,21 +7,68 @@ export class Lazy<T> {
     private _isValueCreated: boolean = false;
     private _value: T | undefined;
 
-    public constructor(private readonly valueFactory: () => T) {
+    public constructor(private readonly valueFactory: () => T, private readonly _valueLifetime?: number) {
     }
 
     public get isValueCreated(): boolean {
         return this._isValueCreated;
     }
 
     public get value(): T {
-        if (!this._isValueCreated) {
-            this._value = this.valueFactory();
-            this._isValueCreated = true;
+        if (this._isValueCreated) {
+            return this._value;
+        }
+
+        this._value = this.valueFactory();
+        this._isValueCreated = true;
+
+        if (this._valueLifetime) {
+            const reset = setTimeout(() => {
+                this._isValueCreated = false;
+                this._value = undefined;
+                clearTimeout(reset);
+            }, this._valueLifetime);
         }
 
         return this._value;
     }
 }
 
-export default Lazy;
+export class AsyncLazy<T> {
+    private _isValueCreated: boolean = false;
+    private _value: T | undefined;
+    private _valuePromise: Promise<T> | undefined;
+
+    public constructor(private readonly valueFactory: () => Promise<T>, private readonly _valueLifetime?: number) {
+    }
+
+    public async getValue(): Promise<T> {
+        if (this._isValueCreated) {
+            return this._value;
+        }
+
+        this._isValueCreated = false;
+
+        const isPrimaryPromise = this._valuePromise === undefined; // The first caller is "primary"
+        const fnLocalPromiseRef = this._valuePromise = this._valuePromise ?? this.valueFactory(); // Await any currently-running Promise if there is one
+        this._value = await fnLocalPromiseRef;
+        this._valuePromise = undefined;
```

Comment:
Are you sure you do not have a race condition here? It seems to me that handlers awaiting on older Promises (if more than one was involved) can reset `this._valuePromise` after a new Promise has been created and is not resolved yet.

Reply:
Once the older one finishes, all of the calls that are awaiting it should finish (since it's the same `Promise`)--so all of them would set `this._valuePromise` to undefined. It's overkill, but I'm not sure which of the executions will start working again first--it may not be the "primary" one.

Reply:
Consider having only the primary promise owner to change `this._value`, `this._valuePromise`, and `this._isValueCreated`. It will be slightly less efficient, but the correctness will be much easier to determine.

The non-primaries would just check if the value is created, and if not, wait on the promise and return the value.

Reply:
Gotcha--so they'd await `this._valuePromise` and return its value, and only primary would set `this._value` / `this._isValueCreated`?

Reply:
Yes, exactly!